source/common/config/config_provider_impl.cc

```diff
@@ -3,29 +3,51 @@
 namespace Envoy {
 namespace Config {
 
-ImmutableConfigProviderImplBase::ImmutableConfigProviderImplBase(
+ImmutableConfigProviderBase::ImmutableConfigProviderBase(
     Server::Configuration::FactoryContext& factory_context,
-    ConfigProviderManagerImplBase& config_provider_manager, ConfigProviderInstanceType type)
+    ConfigProviderManagerImplBase& config_provider_manager,
+    ConfigProviderInstanceType instance_type, ApiType api_type)
     : last_updated_(factory_context.timeSource().systemTime()),
-      config_provider_manager_(config_provider_manager), type_(type) {
+      config_provider_manager_(config_provider_manager), instance_type_(instance_type),
+      api_type_(api_type) {
   config_provider_manager_.bindImmutableConfigProvider(this);
 }
 
-ImmutableConfigProviderImplBase::~ImmutableConfigProviderImplBase() {
+ImmutableConfigProviderBase::~ImmutableConfigProviderBase() {
   config_provider_manager_.unbindImmutableConfigProvider(this);
 }
 
-ConfigSubscriptionInstanceBase::~ConfigSubscriptionInstanceBase() {
+ConfigSubscriptionCommonBase::~ConfigSubscriptionCommonBase() {
   init_target_.ready();
   config_provider_manager_.unbindSubscription(manager_identifier_);
 }
 
-bool ConfigSubscriptionInstanceBase::checkAndApplyConfig(const Protobuf::Message& config_proto,
-                                                         const std::string& config_name,
-                                                         const std::string& version_info) {
+void ConfigSubscriptionCommonBase::bindConfigProvider(MutableConfigProviderCommonBase* provider) {
+  // All config providers bound to a ConfigSubscriptionCommonBase must be of the same concrete
+  // type; this is assumed by ConfigSubscriptionInstance::checkAndApplyConfigUpdate() and is
+  // verified by the assertion below. NOTE: an inlined statement ASSERT() triggers a potentially
+  // evaluated expression warning from clang due to `typeid(**mutable_config_providers_.begin())`.
+  // To avoid this, we use a lambda to separate the first mutable provider dereference from the
+  // typeid() statement.
+  ASSERT([&]() {
```

Comment:
Is there a way to hide this in the underlying `ASSERT`?

Reply:
Sorry, what is 'this' referring to?

Reply:
The lambda complexity. Can the general `ASSERT` just be a lambda?

Reply:
It could be hidden under ASSERT(); I am a tad concerned it would make this code look "unconventional" by having a block of statements as the argument to the macro call though.